File: static/brocard.ts
import isqrt from 'bigint-isqrt';


function sleep(time: number) {
  return new Promise((resolve) => setTimeout(resolve, time));
}


async function factorial(n: bigint): Promise<bigint> {
  let result: bigint = 1n;
  while (n > 1n) {
    await sleep(1);
    result *= n;
    n -= 1n;
  }
  return result;
}

// Math.min does not work on biginters. This does.
const min = (a, b) => (a < b ? a : b);

export async function score(num: bigint): Promise<bigint | string> {
  try {
    if (num < 8) {
      return "Numbers are too small.";
    }
    if (num >= 5000) {
      return "Scoring on server!";
    }
    const nfact = (await factorial(num)) + 1n;
    let smlSqr = isqrt(nfact);
    const bigSqr = (smlSqr + 1n) ** 2n;
    smlSqr = smlSqr ** 2n;
    const interval = bigSqr - smlSqr;
    const sfact = nfact - smlSqr;
    const bfact = bigSqr - nfact;
    const numerator = min(sfact, bfact);
    const bil = BigInt(10 ** 9);
    return (numerator * bil) / interval;
  } catch (e) {
    console.error(e);
    return "Could not score number!";
  }
}

File: app/routers/problems/brocard.py
import asyncio
from math import isqrt
from .utils import parse_integer, register_problem


SLEEP_TIME = 0.1


async def factorial(n):
    result = 1
    while n > 1:
        await asyncio.sleep(SLEEP_TIME)
        result *= n
        n -= 1
    return result


async def score(num: int):
    # Ensure n >= 8
    if num < 8:
        return None
    nfact = await factorial(num) + 1
    sml_sqrt = isqrt(nfact + 1)
    big_sqr = (sml_sqrt + 1) ** 2
    sml_sqr = sml_sqrt ** 2 # Now square!
    interval = big_sqr - sml_sqr
    return int((min(nfact - sml_sqr, big_sqr - nfact) / interval) * 10 ** 9)


register_problem("brocard", score, "Brocard's Problem",
                 "brocard.j2", "lowest", "brocard", parse_integer, 
                 "brocard.svg", submission_type="text")
